File: step_07_recursion_backtracking/7.1_strong_hold/P003_count_good_numbers/solution.py
"""
Problem: Count Good Numbers (LeetCode #1922)
Difficulty: MEDIUM | XP: 25
"""
MOD = 10 ** 9 + 7


# ============================================================
# APPROACH 1: BRUTE FORCE -- Recursion counting all strings
# Time: O(5^(n/2) * 4^(n/2))  |  Space: O(n) stack
#
# Recursively build digit strings: at each even index choose
# from 5 even digits, at each odd index from 4 prime digits.
# Count all valid completions. Only feasible for tiny n.
# ============================================================
def brute_force(n: int) -> int:
    def recurse(pos: int) -> int:
        if pos == n:
            return 1
        choices = 5 if pos % 2 == 0 else 4
        return (choices * recurse(pos + 1)) % MOD

    return recurse(0)


# ============================================================
# APPROACH 2: OPTIMAL -- Math + Fast Modular Exponentiation
# Time: O(log n)  |  Space: O(1)
#
# Even positions: indices 0, 2, 4, ... -> ceil(n/2) positions,
#   each has 5 choices (0,2,4,6,8).
# Odd positions: indices 1, 3, 5, ... -> floor(n/2) positions,
#   each has 4 choices (2,3,5,7).
# Answer = 5^ceil(n/2) * 4^floor(n/2)  (mod 10^9+7)
# Use Python's built-in pow(base, exp, mod) for O(log n).
# ============================================================
def optimal(n: int) -> int:
    even_count = (n + 1) // 2   # ceil(n/2)
    odd_count = n // 2           # floor(n/2)
    return (pow(5, even_count, MOD) * pow(4, odd_count, MOD)) % MOD


# ============================================================
# APPROACH 3: BEST -- Same math, explicit fast-power function
# Time: O(log n)  |  Space: O(1)
#
# Same formula but with a hand-written fast_pow to make the
# binary exponentiation explicit -- useful in Java/C++ interviews
# where built-in modular pow may not be available.
# ============================================================
def best(n: int) -> int:
    def fast_pow(base: int, exp: int, mod: int) -> int:
        result = 1
        base %= mod
        while exp > 0:
            if exp & 1:
                result = result * base % mod
            base = base * base % mod
            exp >>= 1
        return result

    even_count = (n + 1) // 2
    odd_count = n // 2
    return fast_pow(5, even_count, MOD) * fast_pow(4, odd_count, MOD) % MOD


if __name__ == "__main__":
    print("=== Count Good Numbers ===\n")

    tests = [
        (1, 5),
        (4, 400),
        (50, 564908303),
        (3, 100),   # 5*4*5 = 100
        (2, 20),    # 5*4 = 20
    ]

    for n, expected in tests:
        b = brute_force(n) if n <= 20 else "SKIPPED(too large)"
        o = optimal(n)
        bt = best(n)
        ok_o = (o == expected)
        ok_bt = (bt == expected)
        print(f"n={n}  Expected={expected}  Brute={b}  "
              f"Optimal={o}({'OK' if ok_o else 'FAIL'})  "
              f"Best={bt}({'OK' if ok_bt else 'FAIL'})")

File: step_07_recursion_backtracking/7.1_strong_hold/P003_count_good_numbers/Solution.java
/**
 * Problem: Count Good Numbers (LeetCode #1922)
 * Difficulty: MEDIUM | XP: 25
 *
 * @author DSA_Nova
 */
public class Solution {

    private static final long MOD = 1_000_000_007L;

    // ============================================================
    // APPROACH 1: BRUTE FORCE -- Recursion (exponential, tiny n only)
    // Time: O(5^(n/2) * 4^(n/2))  |  Space: O(n) stack
    //
    // At each position decide: even index -> 5 choices, odd -> 4.
    // Multiply choices recursively. Only practical for n <= 15.
    // ============================================================
    static class BruteForce {
        public long countGoodNumbers(long n) {
            return recurse(0, n);
        }

        private long recurse(long pos, long n) {
            if (pos == n) return 1;
            long choices = (pos % 2 == 0) ? 5 : 4;
            return (choices * recurse(pos + 1, n)) % MOD;
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL -- Math + built-in modular exponentiation
    // Time: O(log n)  |  Space: O(1)
    //
    // evenCount = ceil(n/2) positions with 5 choices (0,2,4,6,8)
    // oddCount  = floor(n/2) positions with 4 choices (2,3,5,7)
    // Answer = 5^evenCount * 4^oddCount  (mod 10^9+7)
    // ============================================================
    static class Optimal {
        public long countGoodNumbers(long n) {
            long evenCount = (n + 1) / 2;
            long oddCount  = n / 2;
            return modPow(5, evenCount, MOD) * modPow(4, oddCount, MOD) % MOD;
        }

        private long modPow(long base, long exp, long mod) {
            long result = 1;
            base %= mod;
            while (exp > 0) {
                if ((exp & 1) == 1) result = result * base % mod;
                base = base * base % mod;
                exp >>= 1;
            }
            return result;
        }
    }

    // ============================================================
    // APPROACH 3: BEST -- Same formula, explicit fast-power helper
    // Time: O(log n)  |  Space: O(1)
    //
    // Identical to Optimal; helper is shown explicitly for clarity
    // and portability to languages without built-in modpow.
    // ============================================================
    static class Best {
        public long countGoodNumbers(long n) {
            long evenCount = (n + 1) / 2;
            long oddCount  = n / 2;
            return fastPow(5, evenCount) * fastPow(4, oddCount) % MOD;
        }

        private long fastPow(long base, long exp) {
            long result = 1;
            base %= MOD;
            while (exp > 0) {
                if ((exp & 1L) == 1L) result = result * base % MOD;
                base = base * base % MOD;
                exp >>>= 1;
            }
            return result;
        }
    }

    public static void main(String[] args) {
        System.out.println("=== Count Good Numbers ===\n");

        long[] ns  = {1,  4,   50,        3,   2 };
        long[] exp = {5, 400, 564908303L, 100, 20};

        for (int t = 0; t < ns.length; t++) {
            long n = ns[t];
            // Brute only for small n to avoid stack overflow
            long b  = (n <= 20) ? new BruteForce().countGoodNumbers(n) : -1L;
            long o  = new Optimal().countGoodNumbers(n);
            long bt = new Best().countGoodNumbers(n);
            boolean ok = (o == exp[t] && bt == exp[t]);
            System.out.printf("n=%-12d  Expected=%-12d  Brute=%-12s  Optimal=%-12d  Best=%-12d  [%s]%n",
                    n, exp[t], (b == -1 ? "SKIPPED" : String.valueOf(b)), o, bt, ok ? "OK" : "MISMATCH");
        }
    }
}
